### PR TITLE
verible: init at 0.0-2172-g238b6df6

### PR DESCRIPTION
**Pull request by hzeller**
###### Description of changes

New package: Verible

Verible is a suite of SystemVerilog developer tools, including a parser,
style-linter, formatter, and language server.

https://github.com/chipsalliance/verible

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by viraptor** [comment]
Result of `nixpkgs-review pr 186224` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>verible</li>
  </ul>
</details>

**Comment by viraptor** [comment]
on darwin intel
```
error: hash mismatch in fixed-output derivation '/nix/store/g67y09g67swd962m6ggxlavz7l0ds9f2-verible-v0.0-2172-g238b6df6-deps.tar.gz.drv':
         specified: sha256-XoLdlEeoDJlyWlnXZADHOKu06zKHgHJfgey8UhOt+LM=
            got:    sha256-JjibpdSkz3eiO/PCRx0D7h5u3Q8R44InRyqW9mJSMHo=
error: 1 dependencies of derivation '/nix/store/kl161p2a6c62q8hqybza3rk9vjba3s7v-verible-v0.0-2172-g238b6df6.drv' failed to build
```

**Review by aherrmann — COMMENTED** [review]
LGTM.

I suspect the MacOS issue can be resolved by a dedicated fetchAttrs hash for MacOS. E.g. as done [here](https://github.com/NixOS/nixpkgs/blob/881ea516cf552fbb159aed4462873762a8297409/pkgs/development/python-modules/tensorflow/default.nix#L367-L370).

**Comment by hzeller** [comment]
Interesting that the hash is different on x86 Darwin vs. Linux.
Short of getting a Darwin machine, what are my options to debug this, ideally running on a farm _before_ even creating a pull request ?

In my update I have now limited the platforms to Linux: https://github.com/NixOS/nixpkgs/pull/186224/commits/2c070485a4b2c5edf4abbf589beaf46846b70c1d

... but I'll try @aherrmann 's suggestion to choose the hash depending on the platform.

**Comment by hzeller** [comment]
Question for updates in the future: how can I determine the hash for the fixed output derivation on Darwin when I only have a linux machine ?

**Comment by viraptor** [comment]
No idea for where the difference comes in bazel builds unfortunately. If it's something inherently mac-specific (like case-insensitive filesystem) you could ping the `NixOS/darwin-maintainers` group for an update.

**Comment by aherrmann** [comment]
Bazel repository rules can execute arbitrary code, including executing arbitrary processes and inspecting the file system and environment. `buildBazelPackage` does [some cleanup](https://github.com/NixOS/nixpkgs/blob/881ea516cf552fbb159aed4462873762a8297409/pkgs/build-support/build-bazel-package/default.nix#L107-L112), but there's a good chance that something in the loading phase is still OS dependent. I'm not aware of a reliable method to obtain that hash outside of just running on macOS as well.

**Review by aherrmann — APPROVED** [review]

**Comment by hzeller** [comment]
Since it is approved: Is there anything left to do before someone with permission can merge this PR ?

**Review by ErinvanderVeen — APPROVED** [review]

File: pkgs/development/tools/verible/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/verible/default.nix:44]:
```suggestion
  postPatch = ''
```
Also which files do really need patching?

File: pkgs/development/tools/verible/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/verible/default.nix:75]:
```suggestion
    description = "Suite of SystemVerilog developer tools. Including a style-linter, indexer, formatter, and language server.";
```

File: pkgs/development/tools/verible/default.nix
Comment (hzeller) [reply pkgs/development/tools/verible/default.nix]: Done.
The WORKSPACE file needs to be patched, I now added a comment above in the patch section to clarify why that is needed.

File: pkgs/development/tools/verible/default.nix
Comment (hzeller) [reply pkgs/development/tools/verible/default.nix]: Thanks, that is better. applied.

File: pkgs/development/tools/verible/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/verible/default.nix:48]:
```suggestion
    patchShebangs WORKSPACE
```
if only one file need patching, we should only patch that one. That is a lot faster.

Comment (hzeller) [reply pkgs/development/tools/verible/default.nix:48]: oh, the shebangs need to be patched all over, there are a bunch of bash and python files in here.

I thought you meant the `patches = []` section.

Comment (SuperSandro2000) [reply pkgs/development/tools/verible/default.nix:48]: Do we have a common directory where those files are? We should just avoid patching literally every file.

Comment (hzeller) [reply pkgs/development/tools/verible/default.nix:48]: They are not in a common directory, there are in the order of 45 files in 15 directories involved. Is there a way to narrow the patchShebang pattern with a recursive match of `**/*.py`, `**/*.sh` ?

**Comment by hzeller** [comment]
I've now narrowed down the files to be patched to four, which are explicitly named.

**Comment by hzeller** [comment]
I've also disabled darwin for now until I have a way to test myself what the fixed output derivation hash is.
So this should make this more robust in the first iteration.

**Comment by hzeller** [comment]
Confirmed `nix-build`-ing and working on a Raspberry Pi4 (aarch64-linux).

**Comment by hzeller** [comment]
Anything else needed before this is ready to merge ?

File: pkgs/development/tools/verible/remove-unused-deps.patch
Comment (SuperSandro2000) [inline pkgs/development/tools/verible/remove-unused-deps.patch:24]:
```suggestion
```

File: pkgs/development/tools/verible/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/verible/default.nix:53]:
```suggestion
    patchShebangs bazel/build-version.py \
      common/util/create_version_header.sh \
      common/parser/move_yacc_stack_symbols.sh \
      common/parser/record_syntax_error.sh
```

File: pkgs/development/tools/verible/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/verible/default.nix:29]:
```suggestion
  patches = [
    # Patch WORKSPACE file to not include windows-related dependencies,
    # as they are removed by bazel, breaking the fixed output derivation
    # TODO: fix upstream
    ./remove-unused-deps.patch
  ];
```

File: pkgs/development/tools/verible/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/verible/default.nix:23]:
```suggestion
    rev = version;
```

File: pkgs/development/tools/verible/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/verible/default.nix:18]:
```suggestion
  GIT_VERSION = version;
```

**Review by SuperSandro2000 — COMMENTED** [review]
just some styling

**Comment by hzeller** [comment]
Applied the suggested changes and squashed them all to one commit. (also, whlie at it, rebased to head, as this PR is open for a while)

**Comment by hzeller** [comment]
Anything left to do ?

File: pkgs/development/tools/verible/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/verible/default.nix:23]:
```suggestion
  version = "0.0-2172-g238b6df6";

  # These environment variables are read in bazel/build-version.py to create
  # a build string. Otherwise it would attempt to extract it from .git/.
  GIT_DATE = "2022-08-08";
  GIT_VERSION = "v${version}";

  src = fetchFromGitHub {
    owner = "chipsalliance";
    repo = "verible";
    rev = "v${version}";
```
usually we strip the v prefix which I forgot earlier.

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/development/tools/verible/default.nix
Comment (hzeller) [reply pkgs/development/tools/verible/default.nix]: Done.

**Comment by hzeller** [comment]
Alright, addressed the version number suggestion. Should be ready for merge now.

**Comment by hzeller** [comment]
Ready to merge ?

**Comment by bobby285271** [comment]
(Those approval count labels are updated by bot so sorry if they bring some noise to this PR)

**Pull request merged**
